velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,275 @@
+  auto rawHashes = hashes->asMutable<int64_t>();
+
+  rows.applyToSelected([&](auto row) {
+    int64_t hash = 0;
+    if (!(rawNulls && bits::isBitNull(rawNulls, indices[row]))) {
+      auto size = rawSizes[indices[row]];
+      auto offset = rawOffsets[indices[row]];
+
+      for (int i = 0; i < size; i++) {
+        hash = 31 * hash + rawElementHashes[offset + i];
+      }
+    }
+    rawHashes[row] = hash;
+  });
+}
+
+template <>
+void PrestoHasher::hash<TypeKind::MAP>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  auto baseMap = vector_->base()->as<MapVector>();
+  auto indices = vector_->indices();
+  VELOX_CHECK_EQ(children_.size(), 2)
+
+  auto elementRows = functions::toElementRows(
+      baseMap->mapKeys()->size(), rows, baseMap, indices);
+  BufferPtr keyHashes =
+      AlignedBuffer::allocate<int64_t>(elementRows.end(), baseMap->pool());
+
+  BufferPtr valueHashes =
+      AlignedBuffer::allocate<int64_t>(elementRows.end(), baseMap->pool());
+
+  children_[0]->hash(baseMap->mapKeys(), elementRows, keyHashes);
+  children_[1]->hash(baseMap->mapValues(), elementRows, valueHashes);
+
+  auto rawKeyHashes = keyHashes->as<int64_t>();
+  auto rawValueHashes = valueHashes->as<int64_t>();
+  auto rawHashes = hashes->asMutable<int64_t>();
+
+  auto rawSizes = baseMap->rawSizes();
+  auto rawOffsets = baseMap->rawOffsets();
+  auto rawNulls = baseMap->rawNulls();
+
+  rows.applyToSelected([&](auto row) {
+    int64_t hash = 0;
+    if (!(rawNulls && bits::isBitNull(rawNulls, indices[row]))) {
+      auto size = rawSizes[indices[row]];
+      auto offset = rawOffsets[indices[row]];
+
+      for (int i = 0; i < size; i++) {
+        hash = rawKeyHashes[offset + i] ^ rawValueHashes[offset + i];
+      }
+    }
+    rawHashes[row] = hash;
+  });
+}
+
+template <>
+void PrestoHasher::hash<TypeKind::ROW>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  auto baseRow = vector_->base()->as<RowVector>();
+  auto indices = vector_->indices();
+  SelectivityVector elementRows;
+
+  if (vector_->isIdentityMapping()) {
+    elementRows = rows;
+  } else {
+    elementRows = SelectivityVector(baseRow->size(), false);
+    rows.applyToSelected(
+        [&](auto row) { elementRows.setValid(indices[row], true); });
+    elementRows.updateBounds();
+  }
+
+  BufferPtr childHashes =
+      AlignedBuffer::allocate<int64_t>(elementRows.end(), baseRow->pool());
+
+  auto rawHashes = hashes->asMutable<int64_t>();
+  auto rowChildHashes = childHashes->as<int64_t>();
+
+  std::fill_n(rawHashes, rows.end(), 1);
+
+  for (int i = 0; i < baseRow->childrenSize(); i++) {
+    children_[i]->hash(baseRow->childAt(i), elementRows, childHashes);
+
+    rows.applyToSelected([&](auto row) {
+      rawHashes[row] = 31 * rawHashes[row] + rowChildHashes[indices[row]];
+    });
+  }
+}
+
+void PrestoHasher::hash(
+    const VectorPtr& vector,
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  VELOX_USER_CHECK(
```

Comment:
This should be VELOX_CHECK as this would be a programming error, rather than user error.